internal/services/fingerprint/consumer.go

```diff
@@ -0,0 +1,170 @@
+	consumer := NewConsumer(dbs, i, logger)
+
+	if err := kafka.Consume(ctx, kafka.Config{
+		Brokers: strings.Split(settings.KafkaBrokers, ","),
+		Topic:   settings.DeviceFingerprintTopic,
+		Group:   settings.DeviceFingerprintConsumerGroup,
+	}, consumer.HandleDeviceFingerprint, logger); err != nil {
+		logger.Fatal().Err(err).Msg("couldn't start device fingerprint consumer")
+	}
+
+	logger.Info().Msg("Starting transaction request status listener.")
+
+	return nil
+}
+
+// defaultCredDuration is the default lifetime for VIN credentials. It's meant to cover the
+// current reward week, but contains an extra day for safety.
+var defaultCredDuration = 8 * 24 * time.Hour
+var userDeviceDataPrimaryKeyColumns = []string{models.UserDeviceDatumColumns.UserDeviceID, models.UserDeviceDatumColumns.IntegrationID}
+
+func (c *Consumer) HandleDeviceFingerprint(ctx context.Context, event *Event) error {
+	if !common.IsHexAddress(event.Subject) {
+		return fmt.Errorf("subject %q not a valid address", event.Subject)
+	}
+	addr := common.HexToAddress(event.Subject)
+	signature := common.FromHex(event.Signature)
+	hash := crypto.Keccak256Hash(event.Data)
+
+	if recAddr, err := helpers.Ecrecover(hash.Bytes(), signature); err != nil {
+		return fmt.Errorf("failed to recover an address: %w", err)
+	} else if recAddr != addr {
+		return fmt.Errorf("recovered wrong address %s", recAddr)
+	}
+
+	VIN, err := ExtractVIN(event.Data)
+	if err != nil {
+		if err == ErrNoVIN {
+			return nil
+		}
+		return fmt.Errorf("couldn't extract VIN: %w", err)
+	}
+
+	ad, err := deviceapi.AftermarketDevices(
+		deviceapi.AftermarketDeviceWhere.EthereumAddress.EQ(addr.Bytes()),
+		qm.Load(qm.Rels(deviceapi.AftermarketDeviceRels.VehicleToken, deviceapi.VehicleNFTRels.Claim)),
+	).One(ctx, c.DBS.DBS().Reader)
+	if err != nil {
+		return fmt.Errorf("failed querying for device: %w", err)
+	}
+
+	ud, err := deviceapi.UserDevices(
+		deviceapi.UserDeviceWhere.UserID.EQ(ad.UserID.String),
+	).One(ctx, c.DBS.DBS().Reader)
+
+	if err != nil {
+		return fmt.Errorf("failed querying for device: %w", err)
+	}
+
+	udd, err := models.UserDeviceData(
+		models.UserDeviceDatumWhere.UserDeviceID.EQ(ud.ID),
+	).One(ctx, c.DBS.DBS().Reader)
+
+	if err != nil {
+		return fmt.Errorf("failed querying for device: %w", err)
+	}
+
+	// extract signals with timestamps and persist to signals
+	signals := make(map[string]any)
+	if err := udd.Signals.Unmarshal(&signals); err != nil {
+		return err
+	}
+
+	if vinData, ok := signals["vin"].(map[string]interface{}); ok {
+		if _, exists := vinData["value"]; exists {
+			vinData["value"] = VIN
+		}
+	}
+
+	if err := udd.Signals.Marshal(signals); err != nil {
+		return err
+	}
+
+	if err := udd.Upsert(ctx, c.DBS.DBS().Writer, true, userDeviceDataPrimaryKeyColumns, boil.Infer(), boil.Infer()); err != nil {
+		return fmt.Errorf("error upserting datum: %w", err)
+	}
+
+	c.logger.Info().Str("device-addr", event.Subject).Msg("issued vin credential")
```

Comment:
delete this line, we aren't issuing vin credentials here - no need to report when we update the VIN in user_device_data